### DISCUSSION TITLE
ProjectTo with value converters

### DISCUSSION THREAD

**Discussion by 0x5ECF4ULT** [LuckyPennySoftware/AutoMapper#3652] [NA]
**TL;DR:** There might be a bug when AutoMapper tries to handle `Byte`s and `Char`s.

When building a REST application (not relevant) I ran into a potential bug in AutoMapper. When converting a `byte[]` to `string` without a custom converter, an exception is thrown. I googled a bit around and eventually found out that this behaviour has been removed (#1153). The workaround is to implement two converters (as written below) and upon retry I was greeted with another exception (`System.InvalidOperationException: Missing map from System.Byte to System.Char. Create using Mapper.CreateMap<Byte, Char>`). Well I did that and hoped it would finally work. The result is the stacktrace below.

#### Source type (EFCore mapped model)
```csharp
public partial class Agent
{
    ....
    public byte[] StagingKey { get; set; }
    ....
}
```

#### Destination type
```csharp
public class AgentDto
{
    ....
    public string StagingKey { get; set; }
    ....
}
```

### Mapping configuration

```csharp
public class AgentMapper : Profile
{
    public AgentMapper()
    {
        CreateMap<Agent, AgentDto>()
            .ForMember(dest => dest.StagingKey,
                opt => opt.ConvertUsing<ByteToStringConverter, byte[]>())
            .ReverseMap()
            .ForMember(dest => dest.StagingKey,
                opt => opt.ConvertUsing<StringToByteConverter, string>());
        CreateMap<Byte, Char>();
        CreateMap<Task, TaskDto>().ReverseMap();
    }
}

class ByteToStringConverter : IValueConverter<byte[], string>
{
    public string Convert(byte[] sourceMember, ResolutionContext context) =>
        Encoding.Default.GetString(sourceMember);
}

class StringToByteConverter : IValueConverter<string, byte[]>
{
    public byte[] Convert(string sourceMember, ResolutionContext context) =>
        Encoding.Default.GetBytes(sourceMember);
}
```

### Version: x.y.z
`AutoMapper.Extensions.Microsoft.DependencyInjection`: 8.1.1
`AutoMapper`: 10.1.1
`Microsoft.EntityFrameworkCore.Sqlite`: 5.0.5

### Expected behavior

`string`/`byte[]` objects get converted into their couterpart (eg. `string` -> `byte[]` or the other way around).

### Actual behavior

An exception is thrown (`System.ArgumentException: Argument types do not match`)

For brevity I have only attached the relevant part of the stacktrace.
```
System.ArgumentException: Argument types do not match
   at System.Linq.Expressions.Expression.Bind(MemberInfo member, Expression expression)
   at AutoMapper.QueryableExtensions.Impl.EnumerableExpressionBinder.BindEnumerableExpression(IConfigurationProvider configuration, PropertyMap propertyMap, ExpressionRequest request, ExpressionResolutionResult result, IDictionary`2 typePairCount, LetPropertyMaps letPropertyMaps)
   at AutoMapper.QueryableExtensions.Impl.EnumerableExpressionBinder.Build(IConfigurationProvider configuration, PropertyMap propertyMap, TypeMap propertyTypeMap, ExpressionRequest request, ExpressionResolutionResult result, IDictionary`2 typePairCount, LetPropertyMaps letPropertyMaps)
   at AutoMapper.QueryableExtensions.ExpressionBuilder.<>c__DisplayClass10_0.<CreateMapExpressionCore>g__CreateMemberBinding|4(PropertyExpression propertyExpression, <>c__DisplayClass10_1& )
   at AutoMapper.QueryableExtensions.ExpressionBuilder.<>c__DisplayClass10_0.<CreateMapExpressionCore>g__CreateMemberBindings|1()
   at AutoMapper.QueryableExtensions.ExpressionBuilder.CreateMapExpressionCore(ExpressionRequest request, Expression instanceParameter, IDictionary`2 typePairCount, TypeMap typeMap, LetPropertyMaps letPropertyMaps)
   at AutoMapper.QueryableExtensions.ExpressionBuilder.CreateMapExpressionCore(ExpressionRequest request, Expression instanceParameter, IDictionary`2 typePairCount, LetPropertyMaps letPropertyMaps, TypeMap& typeMap)
   at AutoMapper.QueryableExtensions.ExpressionBuilder.CreateMapExpression(ExpressionRequest request, IDictionary`2 typePairCount, LetPropertyMaps letPropertyMaps)
   at AutoMapper.QueryableExtensions.ExpressionBuilder.CreateMapExpression(ExpressionRequest request)
   at AutoMapper.Internal.LockingConcurrentDictionary`2.<>c__DisplayClass2_1.<.ctor>b__1()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
--- End of stack trace from previous location ---
   at System.Lazy`1.CreateValue()
   at AutoMapper.Internal.LockingConcurrentDictionary`2.GetOrAdd(TKey key)
   at AutoMapper.QueryableExtensions.ExpressionBuilder.GetMapExpression(Type sourceType, Type destinationType, Object parameters, MemberInfo[] membersToExpand)
   at AutoMapper.QueryableExtensions.ProjectionExpression.ToCore(Type destinationType, Object parameters, IEnumerable`1 memberPathsToExpand)
   at AutoMapper.QueryableExtensions.ProjectionExpression.ToCore[TResult](Object parameters, IEnumerable`1 memberPathsToExpand)
   at AutoMapper.QueryableExtensions.ProjectionExpression.To[TResult](Object parameters, Expression`1[] membersToExpand)
   at AutoMapper.QueryableExtensions.Extensions.ProjectTo[TDestination](IQueryable source, IConfigurationProvider configuration, Object parameters, Expression`1[] membersToExpand)
   at AutoMapper.Mapper.ProjectTo[TDestination](IQueryable source, Object parameters, Expression`1[] membersToExpand)
```

### Steps to reproduce

```csharp
[HttpGet]
public async Task<ActionResult<IEnumerable<AgentDto>>> GetAgents()
{
    return await _mapper.ProjectTo<AgentDto>(_context.Agents).ToListAsync();
}
```

**Comment by lbargaoanu** [comment]
https://docs.automapper.org/en/latest/Queryable-Extensions.html#supported-mapping-options
